trackpy/linking.py

```diff
@@ -735,146 +843,262 @@ def link_iter(levels, search_range, memory=0,
+                # add the memory particles to what will be the next source set
+                tmp_set |= self.mem_set
+                # add memory points to prev_hash (to be used as the next source)
+                for m in self.mem_set:
+                    # add points to the hash
+                    prev_hash.add_point(m)
+                    # Record how many times this particle got "held back".
+                    # Since this particle has already been yielded in a previous
+                    # level, we can't store it there. We'll have to put it in the
+                    # track object, then copy this info to the point in cur_hash
+                    # if/when we make a link.
+                    m.track.incr_memory()
+                    # re-create the forward_cands list
+                    m.forward_cands = []
+
+            prev_set = tmp_set
+
+            # TODO: Emit debug message with number of
+            # subnets in this level, numbers of new/remembered/lost particles
+
+            yield cur_level
+
+    def _assign_links(self, dest_set, source_set, search_range,
+                     recursion_counter=0):
+        """Match particles in dest_set with source_set.
+
+        Returns source, dest lists of equal length, corresponding
+        to pairs of source and destination particles. A 'None' value
+        denotes that a match was not found.
+
+        The contents of dest_set and source_set will be changed, as
+        well as the forward_cands and back_cands attributes of the
+        particles. However, this does not meaningfully change the state
+        within link(). All meaningful actions are taken within link(),
+        based on the recommendations of _assign_links().
+        """
+        spl, dpl = [], []
+        diag = self.diag
+        # while there are particles left to link, link
+        while len(dest_set) > 0:
+            p = dest_set.pop()
             bc_c = len(p.back_cands)
             # no backwards candidates
             if bc_c == 0:
-                # add a new track
-                track_lst.append(track_cls(p))
-                # clean up tracking apparatus
-                del p.back_cands
-                # short circuit loop
-                continue
+                # particle will get a new track
+                dpl.append(p)
+                spl.append(None)
+                if diag:
+                    p.diag['search_range'] = search_range
+                continue  # do next dest_set particle
             if bc_c == 1:
                 # one backwards candidate
                 b_c_p = p.back_cands[0]
-                # and only one forward candidatege the repo
+                # and only one forward candidate
                 b_c_p_0 = b_c_p[0]
                 if len(b_c_p_0.forward_cands) == 1:
-                    # add to the track of the candidate
-                    b_c_p_0.track.add_point(p)
-                    _maybe_remove(mem_set, b_c_p_0)
```

Comment:
Where did this memory set removal logic go? (Sorry, I'm struggling to parse the diff here.)